scaling-without-state-transfer.asciidoc

```diff
@@ -0,0 +1,97 @@
+= Scaling up without state transfer
+
+The goal is to be able to add nodes to the cluster and make them own new entries,
+without also owning any of the old entries.
+
+Simplifying assumptions:
+
+* Only one node is being added at a time.
+* Single owner
+* No transactions
+* When scaling down data is just lost
+
+The basic idea is that all keys are written on the newest member.
+The location of all inserted keys is kept in a replicated cache
+(the **anchor cache**) and used for further reads/updates/removals.
+
+It's important to note that clients don't have access to the anchor cache,
+so clients access servers in a round-robin fashion.
+
+
+== Implementation
+
+The implementation will live in a module named `anchored-keys`.
+
+Since we can't deterministically map keys to segments, we don't need a distributed cache.
+Instead we can use an invalidation cache and the `ModuleLifecycle` implementation
+can replace the invalidation interceptor with a custom interceptor.
+
+The interceptor will first look up affected keys in the anchor cache.
+If the key doesn't exist in the anchor cache, reads can assume the value is missing.
+Writes must insert the current writer's location in the anchor cache,
+then forward to the current writer for the actual write.
+
+When a key is removed, the interceptor must also remove the key from the anchor cache.
+When a node leaves, the interceptor must remove all the keys mapped to that node
+from the anchor cache.
+
+== Configuration
+
+Configuration means a custom element `anchored-keys`
+with a single attribute `enabled`.
+
+```
+<invalidation-cache name="name">
+    <anchored-keys enabled="true"/>
```

Comment:
Unless you think there's a need for additional configuration, I think it would be simpler to have this as a boolean attribute on the invalidation-cache element.

```xml
<invalidation-cache name="name" anchored-keys="true" />
```

Reply:
I'd rather put this in a separate module, not in the core, and a module configuration parser needs a root element.

Reply:
:+1: makes sense. I did see you mentioned ModuleLifecycle, but I didn't think of the impact on the config.